File: bindings/dotnet/src/Ccharts/CchartsException.cs
using System.Runtime.InteropServices;

namespace Ccharts;

/// <summary>Status codes reported by the ccharts C ABI.</summary>
public enum CchartsStatus
{
    /// <summary>No error.</summary>
    Ok = 0,
    /// <summary>Empty or mismatched input.</summary>
    InvalidArgument = 1,
    /// <summary>The JSON or CSV document could not be parsed.</summary>
    Parse = 2,
    /// <summary>An allocation failed.</summary>
    OutOfMemory = 3,
    /// <summary>A price was NaN or infinite.</summary>
    NonFinite = 4,
    /// <summary>Width or height was not positive, or exceeded the limits.</summary>
    Dimensions = 5,
}

/// <summary>An error raised by the chart library.</summary>
public sealed class CchartsException : Exception
{
    internal CchartsException(int status)
        : base(Marshal.PtrToStringUTF8(NativeMethods.ErrorMessage(status)) ?? "ccharts error")
    {
        Status = (CchartsStatus)status;
    }

    internal CchartsException(CchartsStatus status, string message) : base(message)
    {
        Status = status;
    }

    /// <summary>The status code behind this error.</summary>
    public CchartsStatus Status { get; }

    internal static void ThrowIfError(int status)
    {
        if (status != 0)
        {
            throw new CchartsException(status);
        }
    }
}

File: bindings/java/src/main/java/io/github/dethrandir/ccharts/CchartsException.java
package io.github.dethrandir.ccharts;

/** An error reported by the ccharts native library. */
public final class CchartsException extends RuntimeException {

    private static final long serialVersionUID = 1L;

    /** Status codes mirroring {@code ccharts_status} in the C ABI. */
    public enum Status {
        /** No error. */
        OK(0),
        /** Empty or mismatched input. */
        INVALID_ARGUMENT(1),
        /** The JSON or CSV document could not be parsed. */
        PARSE(2),
        /** An allocation failed. */
        OUT_OF_MEMORY(3),
        /** A price was NaN or infinite. */
        NON_FINITE(4),
        /** Width or height was not positive, or exceeded the limits. */
        DIMENSIONS(5),
        /** A status code this binding does not know about. */
        UNKNOWN(-1);

        private final int code;

        Status(int code) {
            this.code = code;
        }

        /** The numeric status code. */
        public int code() {
            return code;
        }

        static Status fromCode(int code) {
            for (Status status : values()) {
                if (status.code == code) {
                    return status;
                }
            }
            return UNKNOWN;
        }
    }

    private final Status status;

    CchartsException(Status status, String message) {
        super(message);
        this.status = status;
    }

    CchartsException(int code) {
        this(Status.fromCode(code), Native.errorMessage(code));
    }

    /** The status behind this error. */
    public Status status() {
        return status;
    }

    static void throwIfError(int code) {
        if (code != 0) {
            throw new CchartsException(code);
        }
    }
}
